---
author-meta:
- Benjamin D. Lee
- Alexander J. Titus
- Kun-Hsing Yu
- Marc G. Chevrette
- Paul Allen Stewart
- Evan M. Cofer
- Sebastian Raschka
- Finlay Maguire
- Benjamin J. Lengerich
- Alexandr A. Kalinin
- Anthony Gitter
- Casey S. Greene
- Simina M. Boca
- Timothy J. Triche, Jr.
- Thiago Britto-Borges
- Elana J. Fertig
- Michael D. Kessler
- Alexandra J. Lee
- Beth Signal
bibliography:
- content/manual-references.json
date-meta: '2020-11-29'
header-includes: '<!--

  Manubot generated metadata rendered from header-includes-template.html.

  Suggest improvements at https://github.com/manubot/manubot/blob/master/manubot/process/header-includes-template.html

  -->

  <meta name="dc.format" content="text/html" />

  <meta name="dc.title" content="Ten Quick Tips for Deep Learning in Biology" />

  <meta name="citation_title" content="Ten Quick Tips for Deep Learning in Biology" />

  <meta property="og:title" content="Ten Quick Tips for Deep Learning in Biology" />

  <meta property="twitter:title" content="Ten Quick Tips for Deep Learning in Biology" />

  <meta name="dc.date" content="2020-11-29" />

  <meta name="citation_publication_date" content="2020-11-29" />

  <meta name="dc.language" content="en-US" />

  <meta name="citation_language" content="en-US" />

  <meta name="dc.relation.ispartof" content="Manubot" />

  <meta name="dc.publisher" content="Manubot" />

  <meta name="citation_journal_title" content="Manubot" />

  <meta name="citation_technical_report_institution" content="Manubot" />

  <meta name="citation_author" content="Benjamin D. Lee" />

  <meta name="citation_author_institution" content="Lab41, In-Q-Tel" />

  <meta name="citation_author_institution" content="School of Engineering and Applied Sciences, Harvard University" />

  <meta name="citation_author_institution" content="Department of Genetics, Harvard Medical School" />

  <meta name="citation_author_orcid" content="0000-0002-7133-8397" />

  <meta name="citation_author" content="Alexander J. Titus" />

  <meta name="citation_author_institution" content="Titus Analytics" />

  <meta name="citation_author_orcid" content="0000-0002-0145-9564" />

  <meta name="citation_author" content="Kun-Hsing Yu" />

  <meta name="citation_author_institution" content="Department of Biomedical Informatics, Harvard Medical School" />

  <meta name="citation_author_orcid" content="0000-0001-9892-8218" />

  <meta name="citation_author" content="Marc G. Chevrette" />

  <meta name="citation_author_institution" content="Wisconsin Institute for Discovery and Department of Plant Pathology, University of Wisconsin-Madison" />

  <meta name="citation_author_orcid" content="0000-0002-7209-0717" />

  <meta name="twitter:creator" content="@wildtypeMC" />

  <meta name="citation_author" content="Paul Allen Stewart" />

  <meta name="citation_author_institution" content="Biostatistics and Bioinformatics Shared Resource, Moffitt Cancer Center" />

  <meta name="citation_author_orcid" content="0000-0003-0882-308X" />

  <meta name="citation_author" content="Evan M. Cofer" />

  <meta name="citation_author_institution" content="Lewis-Sigler Institute for Integrative Genomics, Princeton University" />

  <meta name="citation_author_institution" content="Graduate Program in Quantitative and Computational Biology, Princeton University" />

  <meta name="citation_author_orcid" content="0000-0003-3877-0433" />

  <meta name="citation_author" content="Sebastian Raschka" />

  <meta name="citation_author_institution" content="Department of Statistics, University of Wisconsin-Madison" />

  <meta name="citation_author_orcid" content="0000-0001-6989-4493" />

  <meta name="citation_author" content="Finlay Maguire" />

  <meta name="citation_author_institution" content="Faculty of Computer Science, Dalhousie University" />

  <meta name="citation_author_orcid" content="0000-0002-1203-9514" />

  <meta name="citation_author" content="Benjamin J. Lengerich" />

  <meta name="citation_author_institution" content="Computer Science Department, Carnegie Mellon University" />

  <meta name="citation_author_orcid" content="0000-0001-8690-9554" />

  <meta name="citation_author" content="Alexandr A. Kalinin" />

  <meta name="citation_author_institution" content="Department of Computational Medicine and Bioinformatics, University of Michigan" />

  <meta name="citation_author_orcid" content="0000-0003-4563-3226" />

  <meta name="citation_author" content="Anthony Gitter" />

  <meta name="citation_author_institution" content="Department of Biostatistics and Medical Informatics, University of Wisconsin-Madison" />

  <meta name="citation_author_institution" content="Morgridge Institute for Research" />

  <meta name="citation_author_orcid" content="0000-0002-5324-9833" />

  <meta name="twitter:creator" content="@anthonygitter" />

  <meta name="citation_author" content="Casey S. Greene" />

  <meta name="citation_author_institution" content="Department of Systems Pharmacology and Translational Therapeutics, Perelman School of Medicine, University of Pennsylvania" />

  <meta name="citation_author_orcid" content="0000-0001-8713-9213" />

  <meta name="citation_author" content="Simina M. Boca" />

  <meta name="citation_author_institution" content="Innovation Center for Biomedical Informatics, Georgetown University Medical Center" />

  <meta name="citation_author_institution" content="Department of Oncology, Georgetown University Medical Center" />

  <meta name="citation_author_institution" content="Department of Biostatistics, Bioinformatics and Biomathematics, Georgetown University Medical Center" />

  <meta name="citation_author_institution" content="Cancer Prevention and Control Program, Lombardi Comprehensive Cancer Center" />

  <meta name="citation_author_orcid" content="0000-0002-1400-3398" />

  <meta name="citation_author" content="Timothy J. Triche, Jr." />

  <meta name="citation_author_institution" content="Center for Epigenetics, Van Andel Research Institute" />

  <meta name="citation_author_institution" content="Department of Translational Genomics, Keck School of Medicine, University of Southern California" />

  <meta name="citation_author_orcid" content="0000-0001-5665-946X" />

  <meta name="citation_author" content="Thiago Britto-Borges" />

  <meta name="citation_author_institution" content="Section of Bioinformatics and Systems Cardiology, Department of Internal Medicine III and Klaus Tschira Institute for Integrative Computational Cardiology, University Hospital Heidelberg" />

  <meta name="citation_author_orcid" content="0000-0002-6218-4429" />

  <meta name="citation_author" content="Elana J. Fertig" />

  <meta name="citation_author_institution" content="Department of Oncology, Department of Biomedical Engineering, Department of Applied Mathematics and Statistics, Johns Hopkins University" />

  <meta name="citation_author_orcid" content="0000-0003-3204-342X" />

  <meta name="citation_author" content="Michael D. Kessler" />

  <meta name="citation_author_institution" content="Department of Oncology, Johns Hopkins University" />

  <meta name="citation_author_orcid" content="0000-0003-1258-5221" />

  <meta name="citation_author" content="Alexandra J. Lee" />

  <meta name="citation_author_institution" content="Genomics and Computational Biology Graduate Program, University of Pennsylvania" />

  <meta name="citation_author_institution" content="Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania" />

  <meta name="citation_author_orcid" content="0000-0002-0208-3730" />

  <meta name="citation_author" content="Beth Signal" />

  <meta name="citation_author_institution" content="Climate Change Cluster, University of Technology Sydney" />

  <meta name="citation_author_orcid" content="None" />

  <link rel="canonical" href="https://Benjamin-Lee.github.io/deep-rules/" />

  <meta property="og:url" content="https://Benjamin-Lee.github.io/deep-rules/" />

  <meta property="twitter:url" content="https://Benjamin-Lee.github.io/deep-rules/" />

  <meta name="citation_fulltext_html_url" content="https://Benjamin-Lee.github.io/deep-rules/" />

  <meta name="citation_pdf_url" content="https://Benjamin-Lee.github.io/deep-rules/manuscript.pdf" />

  <link rel="alternate" type="application/pdf" href="https://Benjamin-Lee.github.io/deep-rules/manuscript.pdf" />

  <link rel="alternate" type="text/html" href="https://Benjamin-Lee.github.io/deep-rules/v/977cc3ac68364ec7734733be41a8616a900f1596/" />

  <meta name="manubot_html_url_versioned" content="https://Benjamin-Lee.github.io/deep-rules/v/977cc3ac68364ec7734733be41a8616a900f1596/" />

  <meta name="manubot_pdf_url_versioned" content="https://Benjamin-Lee.github.io/deep-rules/v/977cc3ac68364ec7734733be41a8616a900f1596/manuscript.pdf" />

  <meta property="og:type" content="article" />

  <meta property="twitter:card" content="summary_large_image" />

  <meta property="og:image" content="https://github.com/Benjamin-Lee/deep-rules/raw/977cc3ac68364ec7734733be41a8616a900f1596/thumbnail.png" />

  <meta property="twitter:image" content="https://github.com/Benjamin-Lee/deep-rules/raw/977cc3ac68364ec7734733be41a8616a900f1596/thumbnail.png" />

  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />

  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />

  <meta name="theme-color" content="#ad1457" />

  <!-- end Manubot generated metadata -->'
keywords:
- quick tips
- machine learning
- deep learning
- artificial intelligence
lang: en-US
manubot-clear-requests-cache: false
manubot-output-bibliography: output/references.json
manubot-output-citekeys: output/citations.tsv
manubot-requests-cache-path: ci/cache/requests-cache
title: Ten Quick Tips for Deep Learning in Biology
...






<small><em>
This manuscript
([permalink](https://Benjamin-Lee.github.io/deep-rules/v/977cc3ac68364ec7734733be41a8616a900f1596/))
was automatically generated
from [Benjamin-Lee/deep-rules@977cc3a](https://github.com/Benjamin-Lee/deep-rules/tree/977cc3ac68364ec7734733be41a8616a900f1596)
on November 29, 2020.
</em></small>

## Authors
Please note the current author order is chronological and does not reflect the final order.



+ **Benjamin D. Lee**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-7133-8397](https://orcid.org/0000-0002-7133-8397)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [Benjamin-Lee](https://github.com/Benjamin-Lee)<br>
  <small>
     Lab41, In-Q-Tel; School of Engineering and Applied Sciences, Harvard University; Department of Genetics, Harvard Medical School
  </small>

+ **Alexander J. Titus**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-0145-9564](https://orcid.org/0000-0002-0145-9564)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [AlexanderTitus](https://github.com/AlexanderTitus)<br>
  <small>
     Titus Analytics
  </small>

+ **Kun-Hsing Yu**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-9892-8218](https://orcid.org/0000-0001-9892-8218)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [khyu](https://github.com/khyu)<br>
  <small>
     Department of Biomedical Informatics, Harvard Medical School
  </small>

+ **Marc G. Chevrette**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-7209-0717](https://orcid.org/0000-0002-7209-0717)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [chevrm](https://github.com/chevrm)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [wildtypeMC](https://twitter.com/wildtypeMC)<br>
  <small>
     Wisconsin Institute for Discovery and Department of Plant Pathology, University of Wisconsin-Madison
  </small>

+ **Paul Allen Stewart**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-0882-308X](https://orcid.org/0000-0003-0882-308X)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [pstew](https://github.com/pstew)<br>
  <small>
     Biostatistics and Bioinformatics Shared Resource, Moffitt Cancer Center
  </small>

+ **Evan M. Cofer**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-3877-0433](https://orcid.org/0000-0003-3877-0433)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [evancofer](https://github.com/evancofer)<br>
  <small>
     Lewis-Sigler Institute for Integrative Genomics, Princeton University; Graduate Program in Quantitative and Computational Biology, Princeton University
  </small>

+ **Sebastian Raschka**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-6989-4493](https://orcid.org/0000-0001-6989-4493)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [rasbt](https://github.com/rasbt)<br>
  <small>
     Department of Statistics, University of Wisconsin-Madison
  </small>

+ **Finlay Maguire**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-1203-9514](https://orcid.org/0000-0002-1203-9514)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [fmaguire](https://github.com/fmaguire)<br>
  <small>
     Faculty of Computer Science, Dalhousie University
  </small>

+ **Benjamin J. Lengerich**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-8690-9554](https://orcid.org/0000-0001-8690-9554)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [blengerich](https://github.com/blengerich)<br>
  <small>
     Computer Science Department, Carnegie Mellon University
  </small>

+ **Alexandr A. Kalinin**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-4563-3226](https://orcid.org/0000-0003-4563-3226)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [alxndrkalinin](https://github.com/alxndrkalinin)<br>
  <small>
     Department of Computational Medicine and Bioinformatics, University of Michigan
  </small>

+ **Anthony Gitter**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-5324-9833](https://orcid.org/0000-0002-5324-9833)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [agitter](https://github.com/agitter)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [anthonygitter](https://twitter.com/anthonygitter)<br>
  <small>
     Department of Biostatistics and Medical Informatics, University of Wisconsin-Madison; Morgridge Institute for Research
  </small>

+ **Casey S. Greene**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-8713-9213](https://orcid.org/0000-0001-8713-9213)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [cgreene](https://github.com/cgreene)<br>
  <small>
     Department of Systems Pharmacology and Translational Therapeutics, Perelman School of Medicine, University of Pennsylvania
  </small>

+ **Simina M. Boca**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-1400-3398](https://orcid.org/0000-0002-1400-3398)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [SiminaB](https://github.com/SiminaB)<br>
  <small>
     Innovation Center for Biomedical Informatics, Georgetown University Medical Center; Department of Oncology, Georgetown University Medical Center; Department of Biostatistics, Bioinformatics and Biomathematics, Georgetown University Medical Center; Cancer Prevention and Control Program, Lombardi Comprehensive Cancer Center
  </small>

+ **Timothy J. Triche, Jr.**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-5665-946X](https://orcid.org/0000-0001-5665-946X)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [ttriche](https://github.com/ttriche)<br>
  <small>
     Center for Epigenetics, Van Andel Research Institute; Department of Translational Genomics, Keck School of Medicine, University of Southern California
  </small>

+ **Thiago Britto-Borges**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-6218-4429](https://orcid.org/0000-0002-6218-4429)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [tbrittoborges](https://github.com/tbrittoborges)<br>
  <small>
     Section of Bioinformatics and Systems Cardiology, Department of Internal Medicine III and Klaus Tschira Institute for Integrative Computational Cardiology, University Hospital Heidelberg
  </small>

+ **Elana J. Fertig**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-3204-342X](https://orcid.org/0000-0003-3204-342X)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [ejfertig](https://github.com/ejfertig)<br>
  <small>
     Department of Oncology, Department of Biomedical Engineering, Department of Applied Mathematics and Statistics, Johns Hopkins University
  </small>

+ **Michael D. Kessler**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-1258-5221](https://orcid.org/0000-0003-1258-5221)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [mdkessler](https://github.com/mdkessler)<br>
  <small>
     Department of Oncology, Johns Hopkins University
  </small>

+ **Alexandra J. Lee**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-0208-3730](https://orcid.org/0000-0002-0208-3730)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [ajlee21](https://github.com/ajlee21)<br>
  <small>
     Genomics and Computational Biology Graduate Program, University of Pennsylvania; Department of Systems Pharmacology and Translational Therapeutics, University of Pennsylvania
  </small>

+ **Beth Signal**<br>
    · ![GitHub icon](images/github.svg){.inline_icon}
    [betsig](https://github.com/betsig)<br>
  <small>
     Climate Change Cluster, University of Technology Sydney
  </small>



## Introduction {#intro}

Machine learning is a modern approach to problem-solving and task automation. In particular, machine learning is concerned with the development and applications of algorithms that learn how to recognize patterns in data and utilize these for predictive modeling, as opposed to having domain experts developing rules for prediction tasks manually.
Artificial neural networks are a particular class of machine learning algorithms and models that evolved into what we now describe as "deep learning" -- that is, neural networks with many layers (and algorithms that make them perform well).
These neural networks comprise artificial neurons arranged into layers and are modeled after the human brain, even though the building blocks and learning algorithms may differ [@doi:10.1038/s41583-020-0277-3].
Each layer receives input from previous layers (the first of which represents the input data) and then transmits a weighted version of its input to the subsequent layer. 
Thus, the process of "training" a neural network is the tuning of the layers' weights to minimize a 
cost or loss function that serves as a differentiable surrogate of the prediction error.
Deep learning utilizes artificial neural networks with many layers (hence the term "deep"). Given the computational advances made in the last decade, it can now be applied to massive data sets and in innumerable contexts.
In many circumstances, deep learning can learn more complex relationships and make more accurate predictions than other methods.
Therefore, deep learning has become its own subfield of machine learning. In the context of biological research, it has been increasingly used to derive novel insights from high-dimensional biological data [@doi:10.1098/rsif.2017.0387].
For example, deep learning has been used to predict protein-drug binding kinetics [@doi:10.1038/s41467-017-02388-1], to identify the lab-of-origin of synthetic DNA [@doi:10.1038/s41467-018-05378-z], and to uncover the facial phenotypes of genetic disorders [@doi:10.1038/s41591-018-0279-0].

General resources communicating best practices to the scientific community broadly and 
the biological community specifically are scarce, and any resources that do exist are prone 
to reaching obsolescence rapidly due to deep learning's active and specialized nature.
In addition, the lack of established standards or concise recommendations for the application of deep learning to 
biological questions further limits newcomers from using state-of-the-art deep learning in their research.  

To make deep learning more accessible to biological researchers, we solicited input from a community of researchers with varied biological and deep learning interests.
These individuals collaboratively contributed to this manuscript's writing using the GitHub version control platform [@url:https://github.com/Benjamin-Lee/deep-rules] and the Manubot manuscript generation toolset [@url:https://greenelab.github.io/meta-review/].
The goal was to articulate a practical, accessible, and concise set of guidelines and suggestions for biologically oriented researchers to follow when using deep learning.

In the course of our discussions, several themes became clear: the importance of understanding and applying machine learning fundamentals [@doi:10.1186/s13040-017-0155-3] as a baseline for utilizing deep learning, the necessity for extensive model comparisons with careful evaluation, and the need for critical thought in interpreting results generated by deep learning, among others.
The major similarities between deep learning and traditional computational methods also became apparent.
Although deep learning is a distinct subfield of machine learning, it is still a subfield.
It is subject to the many limitations inherent to machine learning, and many best practices for machine learning also apply to deep learning.
In addition, as with all computational methods, deep learning should be applied in a systematic manner that is reproducible and rigorously tested.
Ultimately, the tips we collate range from high-level guidance to best practices for implementation. It is our hope that they will provide actionable, deep learning-specific instruction for both new and experienced deep learning practitioners.
By making deep learning more accessible for use in biological research, we aim to improve the overall usage and reporting quality of deep learning in the literature and enable increasing numbers of researchers to effectively and accurately utilize these state-of-the-art techniques.


## Tip 1: Decide whether deep learning is appropriate for your problem {#appropriate}

In recent years, the number of projects and publications implementing deep learning in biology has risen tremendously [@doi:10.1089/omi.2018.0097; @doi:10.1007/978-981-15-3383-9_54; @doi:10.3390/info11040193]. Given deep learning's usefulness across a range of scientific questions and data modalities, it may seem as though it is a panacea for nearly all modeling problems.
Neural networks that underpin deep learning models are, in fact, universal function approximators and are therefore theoretically capable of learning the functions that relate almost any input and output variables [@doi:10.1007/BF02551274; @tag:hornik-approximation]. However, deep learning is not suited to every modeling situation.
The primary limiting factors for deep learning's suitability for a given problem is primarily limited by the training demands of neural network models, which require significant amounts of data, computing power, and programming as well as modeling expertise.

In the areas of biology where data collection is thoroughly automated, such as DNA sequencing, large amounts of high-quality data may be available.
However, areas of biology that rely on manual data collection may not possess enough data to train and apply deep learning models effectively.
Though there are methods that try to increase the amount of training data, such as data augmentation (in which existing data is slightly manipulated in an attempt to yield "new" samples) and weak supervision (in which simple labeling heuristics are combined to produce noisy, probabilistic labels) [@arxiv:1605.07723v3], these methods cannot overcome substantial data shortages.

In the fields of computer vision and natural language processing, deep neural networks are routinely trained on sample sizes ranging from hundreds of thousands to millions of training examples. Datasets of this size are often not available in many biological contexts. Still, it has been found that under certain circumstances, deep learning can already be considered for datasets with at least one hundred samples per class [@arxiv:1511.06348]. However, it is really best suited for use with datasets that contain orders of magnitude more samples.

Training deep learning models is also very demanding and often requires extensive computing infrastructure and patience to achieve state-of-the-art performance [@doi:10.1109/JPROC.2017.2761740].
In some deep learning contexts, such as generating human-like text, state-of-the-art models have over one hundred billion parameters [@arxiv:2005.14165] and require very costly and time-consuming training procedures [@arxiv:1906.02243].
Though deep learning applications in biology rarely require this much training, they can require computational resources beyond those available on consumer-grade devices such as laptops or office desktops.
Specialized hardware such as discrete graphics processing units (GPUs) and custom deep learning accelerators can dramatically reduce the time and cost required to train models [@doi:10.3390/info11040193]. Still, this hardware is not universally accessible, and cloud-based rentals add additional cost and complexity.
Despite these limiting factors, these specialized hardware solutions are likely to be more broadly available as deep learning becomes more popular (for example, recent-generation iPhones already have such hardware).
In contrast to these large scale computational demands of deep learning, traditional machine learning models can often be trained on laptops (or even on a \$5 computer [@arxiv:1809.00238]) in seconds to minutes.
Therefore, due to this enormous disparity in resource demand alone, traditional machine learning approaches may still prove desirable in various biological applications.

Beyond requiring more data and computational capacity, building and training deep learning models often requires more expertise than training traditional machine learning models.
There are currently several competing programming frameworks for deep learning, such as Tensorflow [@arxiv:1603.04467] and PyTorch [@arxiv:1912.01703], that are widely used across academic research fields and industrial application sets.
While these frameworks allow users to create and deploy entirely novel model architectures, this flexibility combined with the rapid development of the deep learning field has resulted in large and complex frameworks that can be daunting to new users.
For readers new to software development but experienced in biology, gaining computational skills while interfacing with such complex industrial-grade tools can be a prohibitive challenge.
Conversely, traditional machine learning methods are generally more straightforward to implement.
There are currently more tools for automating the model selection and training process for traditional machine learning models than for deep learning models.
For example, automated machine learning (AutoML) tools, such as TPOT [@doi:10.1007/978-3-319-31204-0_9] and Turi Create [@https://github.com/apple/turicreate], are able to test and optimize multiple machine learning models automatically, and can allow users to achieve competitive performance with only a few lines of code.
Thankfully, there are efforts underway to extend these and other automation frameworks to reduce the expertise required to build and use deep learning models.
For example, TPOT, Turi Create, and AutoKeras [@arxiv:1806.10282] are already capable of abstracting away much of the programming required for "standard" deep learning tasks, and high-level interfaces such as Keras [@https://keras.io] and Fastai [@doi:10.3390/info11020108], make it increasingly straightforward to design and test custom deep learning architectures
In the future, projects such as these are likely to make deep learning increasingly accessible to a much wider swatch of researchers.

Despite these limitations, deep learning is strongly indicated over traditional machine learning for specific research questions and problems.
In general, these include problems that feature hidden patterns across the data, complex relationships, and interrelated variables.
Problems in computer vision and natural language processing often exhibit these very features, which helps explain why these areas were some of the first to experience significant breakthroughs during the recent deep learning revolution [@doi:10.1145/3065386]. 
As long as large amounts of accurate and labeled data are available, applications to areas of biology with related data characteristics, such as genetic medicine [@doi:10.3389/fpsyt.2018.00290], radiology [@doi:10.1007/s11604-018-0726-3], microscopy [@doi:10.1364/OPTICA.4.001437], and pharmacovigilance [@doi:10.1093/jamia/ocw180], are similarly likely to benefit from deep learning techniques.
For example, Ferreira et al. used deep learning to recognize individual birds from images [@doi:10.1111/2041-210X.13436] despite this problem being very difficult historically.
By combining automatic data collection using RFID tags with data augmentation and transfer learning (explained in [Tip 5](#architecture)), the authors were able to use deep learning to achieve 90% accuracy across several species.
Another research area where deep learning excels is generative modeling, where new samples are created based on the training data [@doi:10.1109/ISACV.2018.8354080].
One other area of machine learning that has been revolutionized by deep learning is reinforcement learning, which is concerned with training agents to interact with an environment [@arXiv:1709.06560].
Overall, initial evaluation as to whether similar problems (including analogous ones in other domains) have been solved successfully using deep learning can inform researchers about the potential for deep learning to address their needs.

On the other hand, depending on the amount and type of data available and the nature of the problem set, deep learning may not always be able to outperform conventional methods,
As an illustration, Rajkomar et al. [@doi:10.1038/s41746-018-0029-1] found that simpler baseline models achieved performance comparable with deep learning in several clinical prediction tasks using electronic health records.
Another example is provided by Koutsoukas et al., who benchmarked several traditional machine learning approaches against deep neural networks for modeling bioactivity data on moderately sized datasets [@doi:10.1186/s13321-017-0226-y].
The researchers found that while well-tuned deep learning approaches generally tend to outperform conventional classifiers, simple methods such as Naive Bayes classification tend to outperform deep learning as the dataset's noise increases.
Similarly, Chen et al. [@doi:10.1038/s41746-019-0122-0] tested deep learning and a variety of traditional machine learning methods such as logistic regression and random forests on five different clinical datasets.
They found that traditional methods matched or exceeded the accuracy of the deep learning model in all cases despite requiring an order of magnitude less training time.

Therefore, in conclusion, deep learning should only be used after a robust consideration of its strengths and weaknesses for the problem at hand.
Once choosing deep learning as a potential solution, practitioners should still consider traditional methods as performance baselines and use the scientific method to compare the performance of deep learning to that of traditional methods, as outlined in the following tips.


## Tip 2: Use traditional methods to establish performance baselines {#baselines}

Deep learning requires practitioners to consider a larger number and variety of tuning parameters (that is, algorithmic settings) than more traditional machine learning methods.
These settings are often called hyperparameters, and their extensiveness can make it easy to fall into the trap of performing an unnecessarily convoluted analysis.
Hence, before applying deep learning to a given problem, we highly recommend implementing a simpler model with fewer hyperparameters at the beginning of each study.
Such models include logistic regression, random forests, k-nearest neighbors, naive Bayes, and support vector machines, and using them can help to establish baseline performance expectations.
While performance baselines available from existing literature can also serve as helpful guides, an implementation of a simpler model that uses the same software framework as planned for deep learning can greatly help with assessing the correctness of data processing steps, performance evaluation pipelines, resource requirement estimates, and computational performance estimates.
Furthermore, in some cases, it can even be useful to combine simpler baseline models with deep neural networks, as such hybrid models can improve generalization performance, model interpretability, and confidence estimation [@arxiv:1803.04765; @arxiv:1805.11783].

Another potential pitfall arises from comparing the performance of baseline conventional models trained with default settings with the performance of deep learning models that have undergone rigorous tuning and optimization.
Since conventional off-the-shelf machine learning algorithms (for example, support vector machines and random forests) are also likely to benefit from hyperparameter tuning, such incongruity prevents the comparison of equally optimized models and can lead to false conclusions about model efficacy.
Hu and Greene [@doi:10.1142/9789813279827_0033] discuss this under the umbrella of what they call the "Continental Breakfast Included" effect, and they describe how the unequal tuning of hyperparameters across different learning algorithms can especially skew evaluation when the performance of an algorithm varies substantially with modest changes to its hyperparameters.
Therefore, practitioners should tune the settings of both traditional machine and deep learning-based methods before making claims about relative performance differences, as performance comparisons among machine learning and deep learning models are only informative when the models are equally well optimized.

To sum this tip up, practitioners are encouraged to create and fully tune several traditional models and standard pipelines before implementing a deep learning model.


## Tip 3: Understand the complexities of training deep neural networks {#complexities}

Correctly training deep neural networks is a non-trivial process, as there are many different options and potential pitfalls at every stage.
To get good results, you must expect to train many networks across a range of different parameter and hyperparameter settings.
Such training can be made more difficult by the demanding nature of these deep networks, which often require extensive computing infrastructure and optimization in order to achieve state-of-the-art performance [@doi:10.1109/JPROC.2017.2761740].
Furthermore, this experimentation is often noisy, which necessitates increased repetition and exacerbates the organizational challenges inherent to deep learning.
On the whole, all code, random seeds, parameters, and results must be carefully corralled using general good coding practices (e.g. version control [@doi:10.1371/journal.pcbi.1004947], continuous integration etc.) in order to be effective and interpretable.
This organization is also fundamental to the efficient sharing and reproducibility of research work [@doi:10.1371/journal.pcbi.1003285; @arxiv:1810.08055], and to the ability to keep models up to date as new data becomes available.

One specific reproducibility pitfall that is often missed in the application of deep learning is the default use of non-deterministic algorithms by CUDA/CuDNN backends when using GPUs.
That is, the CUDA/CuDNN architectures that facilitate the parallelized computing that power state-of-the-art DL often use algorithms by default that produce different outcomes from iteration to iteration.
Therefore, achieving reproducibility in this context requires explicitly specifying the use of deterministic algorithms (which are typically available within your deep learning library), which is distinct from the setting of random seeds that typically achieve reprocubility by controlling pseudorandom deterministic procedures such as shuffling and initialization [@url:https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#reproducibility].

Similar to the suggestions in [Tip 2](#baselines) about starting with simpler models, try to start with a relatively small network and then increase the size and complexity as needed.
This can help to prevent practitioners from wasting signficant time and resources on running highly complex models that feature numerous unresolved problems. 
Again, beware of the choices that are being made implicitly (that is, by default settings) by your framework of choice (for example, choice of optimization algorithm), as these seemingly trivial specifics can actually have signficant effects on model performance.
For example, adaptive methods often lead to faster convergence during training, but may lead to worse generalization performance on independent datasets [@url:https://papers.nips.cc/paper/7003-the-marginal-value-of-adaptive-gradient-methods-in-machine-learning]).
These nuanced elements are easy to overlook, but it is critical to carefully consider them and to evaluate their potential impact (see [Tip 6](#hyperparameters)).

In short, use smaller and simpler networks to enable faster prototyping and follow general software development best practices to maximize reproducibility.


## Tip 4: Know your data and your question {#know-your-problem}

Having a well defined scientific question and a clear analysis plan is crucial for carrying out a successful deep learning project.
Just like it would be inadvisable to set foot in a laboratory and begin experiments without having a defined endpoint, a deep learning project should not be undertaken without defined goals.
Foremost, it is important to assess if a dataset exists that can answer the biological question of interest using a deep learning-based approach.
If so, obtaining this data (and associated metadata), and reviewing the study protocol, should be pursued as early on in the project as possible.
This can help to ensure that data is as expected and can prevent the wasted time and effort that occur when issues are discovered later on in the analytic process.
For example, a publication or resource might purportedly offer an appropriate dataset that is found to be inadequate upon acquisition.
The data may be unstructured when it is supposed to be structured, crucial metadata such as sample stratification might be missing, or the usable sample size may be different than expected.
Any of these data issues might limit a researcher's ability to use DL to address the biological question at hand, or might otherwise require adjustment before DL can be used.
Data collection should also be carefully documented, or a data collection protocol should be created and specified in the project documentation.
Information about the resources used, download dates, and dataset versions are critical to preserve.
Doing so will help to minimize operational confusion and will increase the reproducibility of the analysis.

Once the dataset is obtained, it is important to learn why and how the data were collected before beginning analysis.
The standardized metadata that exist in many fields can help with this (for example, see [@doi:10.1038/ng1201-365]), but if at all possible, seek out a subject matter expert who has experience with the type of data you are using.
Doing so will minimize guesswork and is likely to increase the success rate of a deep learning project.
For example, one might presume that data collected to test the impact of an intervention derives from a randomized controlled trial.
However, this is not always the case, as ethical or practical concerns often necessitate an observational study design that features prospectively or retrospectively collected data.
In order to ensure similar distributions of important characteristics across study groups in the absence of randomization, such a study may have selected individuals in a fashion that best matches attributes such as age, gender, or weight.
Passively collected datasets can have their own peculiarities, and other study designs can include samples that originate from the same study site, the oversampling of ethnic groups or zip codes, or sample processing differences.
Such information is critical to accurate data analysis, and so it is imperative that practitioners learn about study design assumptions and data specificities prior to performing modeling.
Other study design considerations that should not be overlooked include knowing whether a study involves biological or technical replicates or both.
For example, the existence in a dataset of samples collected from the same individuals at different time points can have significant effects on analyses that make assumptions about sample size and independence (that is, non-independence can lower the effective sample size).
Another potential issue is the existence of systematic biases, which can be induced by confounding variables and can lead to artifacts or so-called "batch effects."
As a consequence, models may learn to rely on the correlations that these systematic biases underpin, even though they are irrelevant to the scientific context of the study.
This can lead to misguided predictions and misleading conclusions [@doi:10.1038/nrg2825].
As described in [Tip 1](#concepts), unsupervised learning and other exploratory analyses can help to identify such biases in these datasets prior to applying a deep learning model.

Overall, practitioners should make sure to thoroughly study their data and understand its context and peculiarities _before_ moving on to performing deep learning.


## Tip 5: Choose an appropriate data representation and neural network architecture {#architecture}

While certain best practices have been established by the research community [@doi:10.1007/978-3-642-35289-8], architecture design choices remain largely problem-specific and are vastly empirical efforts requiring extensive experimentation.
Furthermore, as deep learning is a quickly evolving field, many recommendations are often short-lived, and are frequently replaced by newer insights supported by recent empirical results.
This is further complicated by the fact that many recommendations do not generalize well across different problems and datasets.
Therefore, unfortunately, choosing how to represent your data and design your architecture is closer to an art than a science.
That said, there are some general principles that are useful to follow when experimenting.

First and foremost, use your knowledge of the available data and your question (see [Tip 4](#know-your-problem)) to inform your data representation and architectural design choices.
For example, if your dataset is an array of measurements with no natural ordering of inputs (such as gene expression data), multilayer perceptrons (MLPs) may be effective.
These are the most basic type of neural network, and they are able to learn complex non-linear relationships across the input data despite their relative simplicity.
Similarly, if your dataset is comprised of images, convolutional neural networks (CNNs) are a good choice because they emphasize local structures and adjacency within the data.
CNNs may also be a good choice for learning on sequences, as recent empirical evidence suggests that they can outperform canonical sequence learning techniques such as recurrent neural networks (RNNs) and the closely related long short-term memory (LSTM) networks [@arxiv:1803.01271].

Deep learning models typically benefit from increasing the amount of labeled data with which to train on.
Large amounts of data help to avoid overfitting (see [Tip 7](#overfitting)), and increase the likelihood of achieving top performance on a given task.
In the event that there is not enough data available to train your model, consider using transfer learning.
In transfer learning, a model whose weights were generated by training on another dataset is used as the starting point for training [@tag:Yosinski2014].
Transfer learning is most useful when the pre-training and target datasets are of similar nature [@tag:Yosinski2014].
For this reason, it is important to search for similar datasets that are already available.
These can potentially be used to increase the size of the training set or for pre-training and subsequent fine-tuning on the target data.
However, even when this assumption does not hold, transferring features still can still improve model performance compared with random feature initialization.
For example Rojkomar et al. showed advantages of ImageNet-pretraining [@doi:10.1007/s11263-015-0816-y] for a model that is applied to grayscale medical image classification [@doi:10.1007/s10278-016-9914-9].
In addition, or as an alternative to pre-training models on larger datasets for transfer learning yourself, you may also be able to obtain pre-trained models from public repositories, such as Kipoi [@doi:10.1101/375345] for genomics models.
Moreover, learned features can be helpful even when a pre-training task is different from a target task [@doi:10.1109/CVPRW.2014.131].
Another related approach is multi-task learning, which consists of simultaneously training a network for multiple separate tasks that share features.
In fact, multi-task learning can be used separately or even in combination with transfer learning [@doi:10.1109/TBDATA.2016.2573280].

This tip can be distilled into two main action points: first, base your network's architecture on your knowledge of the problem and, second, take advantage of similar existing data or pre-trained deep learning models.


## Tip 6: Tune your hyperparameters extensively and systematically {#hyperparameters}

Given at least one hidden layer, a non-linear activation function, and a large number of hidden units [@tag:hornik-approximation], multi-layer neural networks can approximate arbitrary continuous functions that relate input and output variables. 
Deeper architectures that feature additonal hidden layers and an increasing number of overall hidden units and learnable weight parameters (the so-called increasing "capacity" of neural networks) allow for solving increaingly complex problems.
However, this increased capacity results in many more parameters to tune, which can pose additional challenges during model training.
In general, one should expect to systematically evaluate the impact of numerous hyperparameters when applying deep neural networks to new data or challenges.
Hyperparameters typically manifest as choices of optimization algorithms, learning rate, activation functions, number of hidden layers and hidden units, size of the training batches, weight initialization schemes, and seeds for pseudo-random number generators used for dataset shuffling and weight initialization.
Moreover, additional hyperparameters are introduced by common techniques that facilitate the training of deeper architectures.
These include norm penalties (typically in the form of $L^2$ regularization), dropout [@tag:srivastava-dropout], and batch normalization [@tag:ioffe-batchnorm], which can reduce the effect of the so-called vanishing or exploding gradient problem when working with deep neural networks.

This wide array of potential parameters can make it difficult to evaluate the extent to which neural network methods are well suited to solving a task, as it can be unclear to practitioners whether previous successful applications were the result of interactions between unique data attributes and specific parameter sets.
Similar to the Continental Breakfast Included effect that we discussed in [Tip 2](#baselines), a lack of clarity on how extensive arrays of hyperparameters were tested and/or chosen can affect methods developers as they attempt to compare techniques.
This effect also has implications for those seeking to use existing deep learning methods, as performance estimates from deep neural networks are often provided after tuning.
The implication of this effect on users of deep neural networks is that attaining performance numbers that match those reported in publications is likely to require signficant effort towards temporally expensive hyperparameter optimization.

Ultimately, to get the best performance of your model, be sure to systematically optimize your hyperparameters on your training dataset, as introduced in the next section.


## Tip 7: Address deep neural networks' increased tendency to overfit the dataset {#overfitting}

Overfitting is a challenge inherent to machine learning in general, and is one of the most significant challenges you'll face when applying deep learning specifically.
Overfitting occurs when a model fits patterns in the training data so closely that it is including non-generalizable noise or non-scientifically relevant perturbations in the relationships it is learning.
In other words, the model fits patterns that are overly specific to the data it is training on rather than learning general relationships that hold across similar datasets.
This subtle distinction is made clearer by seeing what happens when a model is tested on data to which it was not exposed during training: just as a student who memorizes exam materials struggles to correctly answer questions for which they have not studied, a machine learning model that has overfit to its training data will perform poorly on unseen test data.
Deep learning models are particularly susceptible to overfitting due to their relatively large number of parameters and associated representational capacity.
Just as some students may have greater potential for memorization, deep learning models seem more prone to overfitting than machine learning models with fewer parameters.

![A visual example of overfitting and failure to generalize. While a high-degree polynomial achieves high accuracy on its training data, it performs poorly on data with specificities that have not been seen before. That is, the model has learned the training dataset specifically rather than learning a generalizable pattern that represents data of this type. In contrast, a simple linear regression works well on both datasets. The greater representational capacity of the polynomial is analogous to using a larger or deeper neural network.](images/overfitting.png){#fig:overfitting-fig}

In general, one of the most effective ways to combat overfitting is to detect it in the first place.
One way to do this is to split the main dataset being worked on into three independent parts: a training set, a tuning set (also commonly called a validation set in the machine learning literature), and a test set.
These three partitions allow us to optimize models by iterating between model learning on the training set and hyperparameter evaluation on the tuning set without affecting the final model assessment on the test set.
That is, the data used for testing should be "locked away" and used only once to evaluate the final model after all training and tuning steps are completed.
A researcher can then use the model's performance on the independent test data as a measure of how overfit (i.e. non-generalizable) the model is.
This type of approach is necessary for evaluating the generalizability of models without the biases that can arise from learning and testing on the same data [@arxiv:1811.12808; @doi:10.1162/089976698300017197].
While a slight drop in performance from the training set to the test set is normal, a significant drop is a clear sign of overfitting (see Figure @fig:overfitting-fig for a visual demonstration of an overfit model that performs poorly on test data).

If overfitting is an issue, there are a variety of techniques to reduce overfitting, including data augmentation and various regularization techniques [@url:http://jmlr.csail.mit.edu/papers/v15/srivastava14a.html; @tag:krogh-weight-decay].
Another way to reduce overfitting, as described by Chuang and Keiser, is to identify the baseline level of memorization that is occuring by training on data that has its labels randomly shuffled.
By comparing the model performance with the shuffled data to that achieved with the actual data [@doi:10.1021/acschembio.8b00881], a practitioner can identify overfitting as a model that performs no better on real data, as this suggest that any predictive capacity is not due to data-driven signal.
One important caveat when working with partitioned data is the need to apply transformation and normalization procedures equally to all datasets.
The parameters required for such procedures (for example, quantile normalization, a common standardization method when analyzing gene-expression data) should only be derived from the training data, and not from the tuning or test data.
Additionally, many conventional metrics for classification (e.g. area under the receiver operating characteristic curve or AUROC) have limited utility in cases of extreme class imbalance [@pmid:25738806].
Therefore, model performance should be evaluated with a carefully picked panel of relevant metrics that make minimal assumptions about the composition of the testing data [@doi:10.1021/acs.molpharmaceut.7b00578].

When working with biological and medical data, one must also carefully consider potential sources of bias and/or non-independence when defining training and test sets.
For example, a deep learning model for pneumonia detection in chest X-rays appeared to performed well within the hospitals providing the training data, but then failed to generalize to other hospitals [@doi:10.1371/journal.pmed.1002683].
This resulted from the deep learning model picking up on signal related to which hospital the images were from, and represents a type of artifact or "batch effect" that practitioners must be vigilant towards.
When dealing with sequence data, holding out test data that are evolutionarily related or that share structural homology to the training data can result in overfitting that is hard to detect due to the inherent relatedness of the partitioned data (cite?).
In such situations, simply holding out test data selected from a random partition of the training data can be insufficient.
Again, the best remedy for identifying confounding variables is to [know your data](#know-your-problem) and to test models on truly independent data.

In essence, practitioners should split data into training, tuning, and single-use testing sets to assess the performance of the model on data that can provide a reliable estimate of its generalization performance.
Futhermore, be cognizant of the danger of skewed or biased data artificially inflating accuracy.


## Tip 8: Deep learning models can be made more transparent {#blackbox}

While model interpretability is a broad concept, in much of the machine learning literature (including in our guidelines), it refers to the ability to identify the discriminative features that influence or sway the predictions.
In certain cases, the goal behind interpretation is to understand the underlying data generating processes while in other cases the goal is to understand why a model made the prediction that it did for a specific example or set of examples.
Machine learning models vary widely in terms of interpretability: some are fully transparent while others are considered to be "black-boxes" that make predictions with little ability to examine why.
Logistic regression and decision tree models are generally considered interpretable, while deep neural networks are often considered among the most difficult to interpret because they can have many parameters and non-linear relationships. 

Knowing which of the input variables are influencing the outputted predictions, and potentially in what ways, can help with the application or extrapolation of machine learning models.
This is particularly important in biomedicine, where subsequent decision making often requires human input, and where models are employed with the hope of better understanding why relationships exist in the first place.
Furthermore, while prediction rules can be derived from high-throughput molecular datasets, most affordable clinical tests still rely on lower dimensional measurements of a limited number of biomarkers.
Therefore, it is often still unclear how to translate the predictive capacity of deep learning models that encompassing non-linear relationships between countless input variables into clinically digestible terms.
As a result, selecting which biomarkers to use for decision making remains an important modeling and interpretation challenge.
In fact, many authors attribute a lower uptake of deep learning tools in healthcare to interpretability challenges [@doi:10.1109/JBHI.2016.2636665; @doi:10.1038/s41551-018-0315-x]. 
Nonetheless, strategies to interpret both machine learning and deep learning models are rapidly emerging, and the literature on the topic is growing at an exponential rate [@arxiv:2001.02522].
Instead of recommending specific methods for either deep learning-specific or general-purpose model interpretation, we suggest consulting [@url:https://christophm.github.io/interpretable-ml-book/], which is freely available and continually updated.

While active research into model interpretability is enabling increased interpretation of models with many parameters and non-linear relationships, simpler traditional machine learning models often remain substantially easier to interpret.
When deciding on a machine learning approach and model architecture, consider an interpretability versus accuracy tradeoff.
A challenge in considering this tradeoff is that the extent to which one trades interpretability for accuracy depends on the problem itself.
When the features provided to the model are already highly relevant to the task at hand, a simpler and more interpretable model that gives up only a little performance is often more useful.
On the other hand, if features must be combined in complex ways to be meaningful for the task, the performance difference of a model capable of capturing that structure may outweigh the interpretability costs.
An appropriate choice can only be made after careful consideration, which often includes estimating the performance of a simple linear model that serves as a [baseline](#baselines).
In cases where models are learned from high-throughput datasets, a small subset of features in the dataset may be strongly correlated with the complex combination of the larger feature set defined from the deep learning model.
In this case, this more limited number of features can themselves be used in the subsequent simplified model to further enhance interpretability of the model.
This feature reduction can be essential when defining biomarker panels for use in clinical applications.


## Tip 9: Don't over-interpret predictions {#interpretation}

Once we have trained an accurate deep learning model, we often want to use it to deduce relationships and inform scientific findings.
However, in doing this, we need to be careful to correctly interpret the model's predictions.
Given that deep learning models can be difficult to interpret intuitively, there is often a temptation to overinterpret the predictions in indulgent and/or inaccurate ways.
As the classic statistical saying "correlation doesn't mean causation" implies, predictions by deep learning models don't necessarily speak to certain causual relationships.
While we generally know this, and understand that accurately predicting an outcome doesn't imply the learning of any causal mechanism, it can be easy to forget this lesson when the predictions are extremely accurate.
A poignant example of this lesson is from work where authors evaluated the capacities of several models to predict the probability of death for patients with pneumonia admitted to an intensive care unit [@tag:predicting-pneumonia-mortality; @doi:10.1145/2783258.2788613].
Unsurprisingly, the neural network model achieved the best predictive accuracy.
However, after fitting a rule-based model in order to better understand the relationships inherent to their data, the authors discovered that the hospital data implied the rule `HasAsthma(x) => LowerRisk(x)`.
This rule contradicts medical understanding, as having asthma doesn't make pneumonia better!
Nonetheless, this rule was supported by the data, as pneumonia patients with a history of asthma tended to receive more aggressive care.
The neural network had therefore also learned to make predictions according to this rule despite the fact that it has nothing to do with causality or mechanism.
Guiding treatment decisions according to the predictions of the neural network would have been disastrous, even though the neural network had high predictive accuracy.

To trust deep learning models, we must combine knowledge of the training data ([Tip 4](#know-your-problem)) with inspection of the model ([Tip 8](#blackbox)).
To move beyond fitting predictive models and towards the building of an understanding that can inform scientific deduction, we suggest working to disentangle a model's internal logic by comparing data domains where models succeeds to those in which they fail.
By doing so, we can avoid overinterpreting models and view them for what they are: complex statistical models trained on high dimensional data.


## Tip 10: Don't share models trained on sensitive data {#privacy}

Practitioners may encounter datasets that cannot be shared, such as ones for which there would be significant ethical or legal issues associated with their release [@doi:10.1371/journal.pcbi.1005399].
Examples of such data include classified or confidential data, biological data related to trade secrets, medical records, or other personally identifiable information [@doi:10.1038/s41576-020-0257-5].
While deep learning models can capture information-rich abstractions of multiple features of the data during the training process (which represents one of its great strengths), these features may be more prone to leak the data that they were trained over if the model is shared or allowed to be queried with arbitrary inputs [@doi:10.1145/2810103.2813677; @arxiv:1610.05820].
In other words, the complex relationships learned about the input data can potentially be used to infer characteristics about the original dataset.
This means that the strengths that imbue deep learnings with its great predictive capacity also raise the level of risk surrounding data privacy.
Therefore, while there is tremendous promise for deep learning techniques to extract information that cannot readily be captured by traditional methods [@arxiv:1509.09292], it is imperative not to share models trained on sensitive data.
This also holds true for certain traditional machine learning methods that learn by capturing specific details of the full training data (for example, _k_-nearest neighbors models). 

Techniques to train deep neural networks without sharing unencrypted access to data are being advanced through implementations of homomorphic encryption, which serves to enable equivalent prediction on data that is encrypted end to end [@doi:10.1371/journal.pcbi.1006454; @arxiv:1811.00778].
Privacy-preserving techniques [@arxiv:1811.04017], such as differential privacy [@doi:10.1145/2976749.2978318; @doi:10.1101/159756; @arxiv:1812.01484], can help to mitigate risks as long as the assumptions underlying these techniques are met.
These methods provide a path towards a future where trained models and their predictions can be shared, but more software development and theoretical advances will be required to make these techniques easy to apply correctly in many settings.
Unless you use these techniques, don't share the weights or arbitrary access to the predictions of models trained on sensitive data.


## Conclusion {#conclusion}

Deep learning techniques have the potential for wide use in biology and the capacity to meet or exceed the performance of both humans and the current state-of-the art algorithms across a wide range of tasks.
Beyond simply achieving good predictive performance, deep learning has the potential to inform novel biological insights to fundamentally drive high value research.
To realize this potential, the use of deep learning as a research tool must be approached as any other tool would be: scientifically and thoughtfully.
We hope that our tips for applying deep learning to the biological sciences will serve as a starting point for discussion and not as an ending point.


## Acknowledgements {#acknowledgements}

The authors would like the thank Daniel Himmelstein and the developers of Manubot for creating the software that enabled the collaborative composition of this manuscript.
We would also like to thank [**TODO**: insert the names of the contributors who don't meet the standards for authorship] for their contributions to the discussions that comprised the initial stage of the drafting process.


## References {.page_break_before}

<!-- Explicitly insert bibliography here -->
<div id="refs"></div>


[@tag:srivastava-dropout]: http://dl.acm.org/citation.cfm?id=2670313
[@tag:ioffe-batchnorm]: https://dl.acm.org/citation.cfm?id=3045118.3045167
[@tag:garson-interpreting]: https://dl.acm.org/citation.cfm?id=129452
[@tag:hornik-approximation]: doi:10.1016/0893-6080(91)90009-T
[@tag:krogh-weight-decay]: http://dl.acm.org/citation.cfm?id=2986916.2987033
[@tag:predicting-pneumonia-mortality]: doi:10.1016/S0933-3657(96)00367-3
[@tag:Yosinski2014]: https://papers.nips.cc/paper/5347-how-transferable-are-features-in-deep-neural-networks
